### ISSUE TITLE
How to obtain the encrypted sum | mean of a 2D matrix per axis (row-wise and column-wise)?

### ISSUE THREAD

**Issue by SerhatBah** [ibarrond/Pyfhel#152]
Hello, 
and first of all - Thank you very very much for your help. Everything you say is incredibly helpful and means a lot to me!

I have implemented everything in our code, as you said. Now I am struggling with other error messages:

**CODE**

```
HE = Pyfhel()  # Creating empty Pyfhel object
ckks_params = {
	'scheme': 'CKKS',  # can also be 'ckks'
	'n': 2 ** 14,  # Polynomial modulus degree. For CKKS, n/2 values can be
	#  encoded in a single ciphertext.
	#  Typ. 2^D for D in [10, 16]
	'scale': 2 ** 30,  # All the encodings will use it for float->fixed point
	#  conversion: x_fix = round(x_float * scale)
	#  You can use this as default scale or use a different
	#  scale on each operation (set in HE.encryptFrac)
	'qi_sizes': [60, 30, 30, 30, 60]  # Number of bits of each prime in the chain.
	# Intermediate values should be  close to log2(scale)
	# for each operation, to have small rounding errors.
}

HE.contextGen(**ckks_params)  # Generate context for bfv scheme
HE.keyGen()  # Key Generation: generates a pair of public/secret keys
HE.rotateKeyGen()

data = np.random.randint(0, 5, size=(10, 5))
num_rows, num_cols = data.shape
rows = np.ones(num_rows, dtype=np.bool)
cols = np.ones(num_cols, dtype=np.bool)

 # Encrypting sub_data
        # 1. make sub_data a contiguous array in memory
        # 2. change 2d arrays into 1d
        # 3. Convert plaintext into ciphertext

		
        enc_sub_data = sub_data.flatten()
        #arr_sub_data = np.empty(len(enc_sub_data), dtype=PyCtxt)
        sub_data = data[rows][:, cols]
        sub_data = np.ascontiguousarray(sub_data).flatten().astype(np.float64)
        arr_sub_data = HE.encryptFrac(sub_data)
```

**ERROR**

```
/usr/bin/python3.10 /home/serhat/SeCCA-CKKS/secured_cheng_church_yeast.py 
SeCCA Step 2
Number of the Bicluster:1
Traceback (most recent call last):
  File "/home/serhat/SeCCA-CKKS/secured_cheng_church_yeast.py", line 20, in <module>
    biclustering = secca.run(data)
  File "/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py", line 120, in run
    self._multiple_node_deletion(data, rows, cols, msr_thr, HE, t_enc, t_dec)
  File "/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py", line 167, in _multiple_node_deletion
    msr, row_msr, col_msr = self._calculate_msr(data, rows, cols, HE, t_enc, t_dec)
  File "/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py", line 240, in _calculate_msr
    enc_row_means = np.mean(arr_sub_data, axis=1)
  File "<__array_function__ internals>", line 5, in mean
  File "/home/serhat/.local/lib/python3.10/site-packages/numpy/core/fromnumeric.py", line 3440, in mean
    return _methods._mean(a, axis=axis, dtype=dtype,
  File "/home/serhat/.local/lib/python3.10/site-packages/numpy/core/_methods.py", line 167, in _mean
    rcount = _count_reduce_items(arr, axis, keepdims=keepdims, where=where)
  File "/home/serhat/.local/lib/python3.10/site-packages/numpy/core/_methods.py", line 76, in _count_reduce_items
    items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
numpy.AxisError: axis 1 is out of bounds for array of dimension 0

Process finished with exit code 1
```

We should not use reshape anymore because we are not encrypting each element of the array individually. But now we can't get the mean of either axis =0 or axis =1 because we don't have our dimensions anymore. Would you please help us with this as well?

**Comment by ibarrond** [comment]
Hi @SerhatBah , it seems your error is not present in the code you wrote here! I can execute the code you posted without  any issue! please add the code that is generating an error.

Let me rephrase the question then: **How to obtain the encrypted mean per axis in a 2D matrix?**

Here is the idea

```python
import numpy as np
from Pyfhel import Pyfhel, PyCtxt

# HE SETUP
HE = Pyfhel(context_params={'scheme': 'CKKS', 'n': 2 ** 14,
                            'scale': 2 ** 30, 'qi_sizes': [60, 30, 30, 30, 60]},
            key_gen=True)
HE.rotateKeyGen()     # We will use it to rotate the values in the vector

# DATA GENERATION
n_rows, n_cols = 10, 5
n_elements = n_rows*n_cols
np.random.seed(42)                                    # Fixed seed for reproducibility
data = np.random.randint(0, 5, size=(n_rows, n_cols)) # Generate data

# ------------------------------------------------------------------------------
# NAIVE METHOD (clocks 310ms on my laptop)
# Encryption -> Flatten to 1D vector of size (n_rows x n_cols) and encrypt as-is
flattened_data = data.flatten()                       
enc_data = HE.encryptFrac(flattened_data.astype(np.float64))

# Row-wise sum
enc_rowwise_sum = PyCtxt(copy_ctxt=enc_data)
for i in range(1,n_cols):
    enc_rowwise_sum += enc_data << i
# Column-wise sum
enc_colwise_sum = PyCtxt(copy_ctxt=enc_data)
for i in range(1, n_rows):
    enc_colwise_sum += enc_data << (i*n_cols)

# Decrypt, select only the slots with results and round their values
res_rowsum = enc_rowwise_sum.decrypt()[:n_elements:n_cols].round()
res_colsum = enc_colwise_sum.decrypt()[:n_cols].round()


# ------------------------------------------------------------------------------
# OPTIMIZED METHOD
# Encryption -> Flatten by aligning to a power of 2 in the first axis, then encrypt. Rotations in powers of 2 are the cheapest!
#   Size: (n_rows x ceil(log2(n_cols)))
n_zero_cols       = int(np.ceil(np.log2(n_cols)))
n_cols_filled     = n_zero_cols + n_cols
n_elements_filled = n_cols_filled * n_rows

trailing_zeros = np.zeros(shape=(n_rows, n_zero_cols))
filled_data    = np.concatenate([data, trailing_zeros], axis=1)
flattened_data = filled_data.flatten()                       
enc_data       = HE.encryptFrac(flattened_data.astype(np.float64))

# Row-wise sum (clocks 100ms on my laptop)
enc_rowwise_sum_opt = PyCtxt(copy_ctxt=enc_data)
for i in range(int(np.log2(n_cols_filled))):
    enc_rowwise_sum_opt += enc_rowwise_sum_opt << (2**i)
    
# Column-wise sum
enc_colwise_sum_opt = PyCtxt(copy_ctxt=enc_data)
for j in range(int(np.ceil(np.log2(n_rows)))):
    enc_colwise_sum_opt += enc_colwise_sum_opt << ((2**j)*n_cols_filled)

res_rowsum_opt = enc_rowwise_sum_opt.decrypt()[:n_elements_filled:n_cols_filled].round()
res_colsum_opt = enc_colwise_sum_opt.decrypt()[:n_cols].round()

# ------------------------------------------------------------------------------
# VERIFICATION
assert np.allclose(data.sum(axis=0), res_colsum)
assert np.allclose(data.sum(axis=1), res_rowsum)
assert np.allclose(data.sum(axis=0), res_colsum_opt)
assert np.allclose(data.sum(axis=1), res_rowsum_opt)

**Comment by ibarrond** [comment]
Closing as completed. Feel free to reopen if you still have doubts.

**Comment by ibarrond** [comment]
You no longer need the `for` loop!

```python
enc_rowwise_mean = enc_rowwise_sum / n_cols
```

**Comment by SerhatBah** [comment]
Hello,
Thank you for the help and the code fragment. That really really helped a lot.
I have tried to implement as much of it as possible. To show that, I am submitting my code fragment here:

```
HE = Pyfhel()  # Creating empty Pyfhel object
ckks_params = {
	'scheme': 'CKKS',  # can also be 'ckks'
	'n': 2 ** 14,  # Polynomial modulus degree. For CKKS, n/2 values can be
	#  encoded in a single ciphertext.
	#  Typ. 2^D for D in [10, 16]
	'scale': 2 ** 30,  # All the encodings will use it for float->fixed point
	#  conversion: x_fix = round(x_float * scale)
	#  You can use this as default scale or use a different
	#  scale on each operation (set in HE.encryptFrac)
	'qi_sizes': [60, 30, 30, 30, 60]  # Number of bits of each prime in the chain.
	# Intermediate values should be  close to log2(scale)
	# for each operation, to have small rounding errors.
}

HE.contextGen(**ckks_params)  # Generate context for bfv scheme
HE.keyGen()  # Key Generation: generates a pair of public/secret keys
HE.rotateKeyGen()
HE.relinKeyGen()  # Relinearization key generation


data = np.random.randint(0, 5, size=(10, 5))
num_rows, num_cols = data.shape
rows = np.ones(num_rows, dtype=np.bool)
cols = np.ones(num_cols, dtype=np.bool)

 # Encrypting sub_data
        # 1. make sub_data a contiguous array in memory
        # 2. change 2d arrays into 1d
        # 3. Convert plaintext into ciphertext

		
        enc_sub_data = sub_data.flatten()
        #arr_sub_data = np.empty(len(enc_sub_data), dtype=PyCtxt)
        sub_data = data[rows][:, cols]
        sub_data = np.ascontiguousarray(sub_data).flatten().astype(np.float64)
        arr_sub_data = HE.encryptFrac(sub_data)

        # Encrypting sub_data
        # 1. make sub_data a contiguous array in memory
        # 2. change 2d arrays into 1d
        # 3. Convert plaintext into ciphertext


        sub_data = data[rows][:, cols]
        sub_data = np.ascontiguousarray(sub_data).flatten().astype(np.float64)
        arr_sub_data = HE.encryptFrac(sub_data)
        n_elements = len(rows) * len(cols)

        # Row-wise sum & Encrypting row_means
        enc_rowwise_sum = PyCtxt(copy_ctxt=arr_sub_data)
        for i in range(1, len(cols)):
            enc_rowwise_sum += arr_sub_data << i

        # Column-wise sum & Encrypting col_means
        enc_colwise_sum = PyCtxt(copy_ctxt=arr_sub_data)
        for i in range(1, len(rows)):
            enc_colwise_sum += arr_sub_data << (i * len(cols))

        # Encrypting data_mean
        enc_data_mean = np.sum(arr_sub_data) / len(arr_sub_data)

        # 1. Mean
        enc_row_means = enc_rowwise_sum / len(cols)
        enc_col_means = enc_colwise_sum / len(rows)

        # Encrypting Residues
        enc_residues = arr_sub_data - enc_row_means - enc_col_means + enc_data_mean

        # Encrypting Squared Residues
        enc_squared_residues = enc_residues ** 2

        # Encrypting msr
        enc_msr = np.sum(enc_squared_residues) / len(enc_squared_residues)

        # Encrypting row_msr
        enc_row_msr = enc_squared_residues / len(enc_squared_residues)

        # Encrypting col_msr
        enc_col_msr = enc_squared_residues / len(enc_squared_residues)

        # Decrypting msr
        t_dec0 = time.perf_counter()
        decrypted_msr = enc_squared_residues.decrypt()[:n_elements:len(enc_squared_residues)].round()

        # Decrypting msr_row
        decrypted_msr_row = enc_rowwise_sum.decrypt()[:n_elements:len(cols)].round()

        # Decrypting msr_col
        decrypted_msr_col = enc_colwise_sum.decrypt()[:len(cols)].round()

        return decrypted_msr, decrypted_msr_row, decrypted_msr_col
```



Now I am faced with the task of obtaining the Encrypting Mean Squared Residues (msr, row_msr, col_msr) in this section:


 ```
       # Encrypting msr
        enc_msr = np.sum(enc_squared_residues) / len(enc_squared_residues)

        # Encrypting row_msr
        enc_row_msr = enc_squared_residues / len(enc_squared_residues)

        # Encrypting col_msr
        enc_col_msr = enc_squared_residues / len(enc_squared_residues)
```

But for this I get this error message: 

```
/usr/bin/python3.10 /home/serhat/SeCCA-CKKS/secured_cheng_church_yeast.py 
SeCCA Step 2
Number of the Bicluster:5
/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py:258: UserWarning: Cannot align scales 120 and 30 (available rescalings: [30 60])
  enc_msr = np.sum(enc_squared_residues) / len(enc_squared_residues)
Traceback (most recent call last):
  File "/home/serhat/SeCCA-CKKS/secured_cheng_church_yeast.py", line 27, in <module>
    biclustering = secca.run(data)
  File "/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py", line 123, in run
    self._multiple_node_deletion(data, rows, cols, msr_thr, HE, t_enc, t_dec)
  File "/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py", line 170, in _multiple_node_deletion
    msr, row_msr, col_msr = self._calculate_msr(data, rows, cols, HE, t_enc, t_dec)
  File "/home/serhat/SeCCA-CKKS/biclustlib/algorithms/type2.py", line 258, in _calculate_msr
    enc_msr = np.sum(enc_squared_residues) / len(enc_squared_residues)
  File "Pyfhel/PyCtxt.pyx", line 556, in Pyfhel.PyCtxt.PyCtxt.__truediv__
  File "Pyfhel/Pyfhel.pyx", line 1107, in Pyfhel.Pyfhel.Pyfhel.multiply_plain
  File "Pyfhel/Pyfhel.pyx", line 1126, in Pyfhel.Pyfhel.Pyfhel.multiply_plain
ValueError: scale out of bounds
corrupted double-linked list

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```


The assumption is that due to the large multiplication the ciphertext is too large and I now have to apply the rescale-function.

I have looked at the CKKS_Demo file and know that there is this command: `HE.rescale_to_next(ctxt) `

But I can't find a way to implement it in my code, can you help me please? Or is the problem something else? I would be very pleased to receive an answer from you, as always..:)
